File: github_analyzer.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

require 'net/http'
require 'json'
require 'optparse'
require 'time'

##
# GitHub Analyzer - Ruby Implementation
# A elegant and idiomatic Ruby approach to analyzing GitHub repositories
#

class GitHubAnalyzer
  BASE_URL = 'https://api.github.com'
  DEFAULT_TIMEOUT = 30
  
  attr_reader :token, :http_client

  def initialize(token = nil)
    @token = token || ENV['GITHUB_TOKEN']
    @http_client = Net::HTTP
  end

  ##
  # Fetch repository information from GitHub API
  #
  def get_repository(owner, repo)
    endpoint = "/repos/#{owner}/#{repo}"
    response = make_request(endpoint)
    
    RepositoryInfo.new(
      name: response['name'],
      description: response['description'],
      stars: response['stargazers_count'],
      forks: response['forks_count'],
      watchers: response['watchers_count'],
      language: response['language'],
      topics: response['topics'] || [],
      created_at: response['created_at'],
      updated_at: response['updated_at'],
      open_issues: response['open_issues_count'],
      url: response['html_url'],
      is_archived: response['archived'],
      is_fork: response['fork']
    )
  end

  ##
  # Fetch repository contributors
  #
  def get_contributors(owner, repo)
    endpoint = "/repos/#{owner}/#{repo}/contributors?per_page=100"
    response = make_request(endpoint)
    
    response.map do |contributor|
      Contributor.new(
        login: contributor['login'],
        contributions: contributor['contributions'],
        avatar_url: contributor['avatar_url']
      )
    end
  end

  ##
  # Fetch programming language distribution
  #
  def get_languages(owner, repo)
    endpoint = "/repos/#{owner}/#{repo}/languages"
    response = make_request(endpoint)
    response || {}
  end

  ##
  # Perform complete analysis of a repository
  #
  def analyze(owner, repo)
    start_time = Time.now
    
    # Fetch all data concurrently (simulated with sequential calls)
    repository = get_repository(owner, repo)
    contributors = get_contributors(owner, repo)
    languages = get_languages(owner, repo)
    
    health_score = calculate_health(repository, contributors)
    duration = Time.now - start_time
    
    AnalysisResult.new(
      repository: repository,
      contributors: contributors,
      languages: languages,
      health: health_score,
      duration: duration
    )
  rescue StandardError => e
    raise "Analysis failed: #{e.message}"
  end

  private

  ##
  # Make HTTP request to GitHub API
  #
  def make_request(endpoint)
    uri = URI("#{BASE_URL}#{endpoint}")
    
    http = @http_client.new(uri.host, uri.port)
    http.use_ssl = true
    http.read_timeout = DEFAULT_TIMEOUT
    
    request = Net::HTTP::Get.new(uri.request_uri)
    request['Accept'] = 'application/vnd.github.v3+json'
    request['Authorization'] = "token #{@token}" if @token
    
    response = http.request(request)
    
    case response.code.to_i
    when 200
      JSON.parse(response.body)
    when 404
      raise "Repository not found"
    when 401
      raise "Unauthorized - check your GitHub token"
    when 403
      raise "Rate limit exceeded"
    else
      raise "HTTP #{response.code}: #{response.body}"
    end
  end

  ##
  # Calculate health score for repository
  #
  def calculate_health(repository, contributors)
    score = 0
    insights = []

    # Stars evaluation (0-25)
    case repository.stars
    when 1000..Float::INFINITY
      score += 25
      insights << "🌟 Excellent community engagement"
    when 500...1000
      score += 20
      insights << "⭐ Good community interest"
    when 100...500
      score += 15
      insights << "📈 Growing community"
    else
      insights << "🚀 Building community awareness"
    end

    # Contributors evaluation (0-25)
    case contributors.length
    when 50..Float::INFINITY
      score += 25
      insights << "👥 Strong contributor base"
    when 20...50
      score += 20
      insights << "🤝 Active community"
    when 5...20
      score += 15
      insights << "👤 Growing team"
    else
      insights << "👤 Building contributor network"
    end

    # Forks evaluation (0-20)
    if repository.stars > 0
      fork_ratio = repository.forks.to_f / repository.stars
      case fork_ratio
      when 0.3..Float::INFINITY
        score += 20
        insights << "🔀 High fork-to-star ratio"
      when 0.1...0.3
        score += 15
        insights << "🔀 Moderate fork activity"
      else
        score += 10
      end
    end

    # Issues evaluation (0-15)
    case repository.open_issues
    when 0...20
      score += 15
      insights << "✅ Well-maintained issue tracker"
    when 20...50
      score += 10
      insights << "📋 Manageable issue backlog"
    else
      insights << "⚠️ Consider addressing issue backlog"
    end

    # Activity evaluation (0-15)
    updated_at = Time.parse(repository.updated_at)
    days_since_update = (Time.now - updated_at) / 86400.0

    case days_since_update
    when 0...7
      score += 15
      insights << "🔥 Actively maintained"
    when 7...30
      score += 10
      insights << "✓ Regularly maintained"
    when 30...90
      score += 5
      insights << "⏱️ Maintenance may be slowing"
    end

    rating = case score
             when 90..100 then "Excellent"
             when 70...90 then "Good"
             when 50...70 then "Fair"
             else "Needs Improvement"
             end

    HealthScore.new(score: score, rating: rating, insights: insights)
  end
end

##
# Value object for repository information
#
class RepositoryInfo
  attr_reader :name, :description, :stars, :forks, :watchers,
              :language, :topics, :created_at, :updated_at,
              :open_issues, :url, :is_archived, :is_fork

  def initialize(**attrs)
    attrs.each { |key, value| instance_variable_set("@#{key}", value) }
  end
end

##
# Value object for contributor information
#
class Contributor
  attr_reader :login, :contributions, :avatar_url

  def initialize(login:, contributions:, avatar_url:)
    @login = login
    @contributions = contributions
    @avatar_url = avatar_url
  end
end

##
# Value object for health score
#
class HealthScore
  attr_reader :score, :rating, :insights

  def initialize(score:, rating:, insights:)
    @score = score
    @rating = rating
    @insights = insights
  end
end

##
# Value object for analysis result
#
class AnalysisResult
  attr_reader :repository, :contributors, :languages, :health, :duration

  def initialize(repository:, contributors:, languages:, health:, duration:)
    @repository = repository
    @contributors = contributors
    @languages = languages
    @health = health
    @duration = duration
  end

  ##
  # Pretty print analysis results
  #
  def print_report
    puts "\n╔════════════════════════════════════════════════════════╗"
    puts "║           GITHUB REPOSITORY ANALYSIS REPORT            ║"
    puts "╚════════════════════════════════════════════════════════╝\n"

    puts "📦 Repository: #{repository.name}"
    puts "📝 Description: #{repository.description}" if repository.description

    puts "\n📊 Statistics:"
    puts "  ⭐ Stars: #{repository.stars}"
    puts "  🔀 Forks: #{repository.forks}"
    puts "  👁️ Watchers: #{repository.watchers}"
    puts "  ⚠️ Open Issues: #{repository.open_issues}"

    puts "\n👥 Top Contributors: #{contributors.length} total"
    contributors.take(5).each_with_index do |contributor, index|
      puts "  #{index + 1}. #{contributor.login} (#{contributor.contributions} commits)"
    end

    print_languages

    print_health

    puts "\n⏱️ Analysis completed in #{(duration * 1000).round(2)}ms\n"
  end

  private

  def print_languages
    return if languages.empty?

    puts "\n💻 Languages:"
    total_bytes = languages.values.sum
    languages.sort_by { |_, bytes| -bytes }.take(5).each do |lang, bytes|
      percent = ((bytes.to_f / total_bytes) * 100).round(1)
      puts "  #{lang}: #{percent}%"
    end
  end

  def print_health
    puts "\n🏥 Health Assessment:"
    puts "  Health Score: #{health.score}/100"
    puts "  Rating: #{health.rating}"
    puts "\n  Insights:"
    health.insights.each { |insight| puts "  #{insight}" }
  end
end

##
# CLI interface
#
class CLI
  def self.run
    options = {}

    OptionParser.new do |opts|
      opts.banner = "GitHub Analyzer - Ruby Version\nUsage: ruby github_analyzer.rb <owner/repo> [options]"

      opts.on('-t', '--token TOKEN', 'GitHub API token') do |token|
        options[:token] = token
      end

      opts.on('-h', '--help', 'Show this help message') do
        puts opts
        exit
      end
    end.parse!

    if ARGV.empty?
      puts "Usage: ruby github_analyzer.rb <owner/repo> [--token TOKEN]"
      puts "Set GITHUB_TOKEN environment variable to avoid using --token flag"
      exit 1
    end

    repo_path = ARGV[0]
    owner, repo = repo_path.split('/')

    unless owner && repo
      puts "❌ Error: Invalid repository format. Use: owner/repo"
      exit 1
    end

    analyzer = GitHubAnalyzer.new(options[:token])
    puts "\n📊 Analyzing #{owner}/#{repo}...\n"

    result = analyzer.analyze(owner, repo)
    result.print_report
  rescue StandardError => e
    puts "❌ Error: #{e.message}"
    exit 1
  end
end

# Run CLI if executed directly
CLI.run if __FILE__ == $PROGRAM_NAME

File: github_analyzer.ts
#!/usr/bin/env node
/**
 * GitHub Analyzer CLI - TypeScript Version
 * A modern TypeScript implementation for analyzing GitHub repositories
 * with type safety and concurrent request handling
 */

import axios, { AxiosInstance } from 'axios';
import { performance } from 'perf_hooks';

interface GitHubRepo {
  name: string;
  description: string | null;
  stars: number;
  forks: number;
  watchers: number;
  language: string | null;
  topics: string[];
  createdAt: string;
  updatedAt: string;
  openIssues: number;
}

interface Contributor {
  login: string;
  contributions: number;
  avatarUrl: string;
}

interface AnalysisResult {
  repository: GitHubRepo;
  contributors: Contributor[];
  languages: { [key: string]: number };
  health: RepositoryHealth;
  analysisTime: number;
}

interface RepositoryHealth {
  score: number;
  rating: 'Excellent' | 'Good' | 'Fair' | 'Needs Improvement';
  insights: string[];
}

class GitHubAnalyzerTS {
  private client: AxiosInstance;
  private token: string | null;

  constructor(token?: string) {
    this.token = token || process.env.GITHUB_TOKEN || null;
    this.client = axios.create({
      baseURL: 'https://api.github.com',
      headers: {
        'Accept': 'application/vnd.github.v3+json',
        ...(this.token && { 'Authorization': `token ${this.token}` })
      }
    });
  }

  async getRepositoryInfo(owner: string, repo: string): Promise<GitHubRepo> {
    try {
      const response = await this.client.get(`/repos/${owner}/${repo}`);
      const data = response.data;

      return {
        name: data.name,
        description: data.description,
        stars: data.stargazers_count,
        forks: data.forks_count,
        watchers: data.watchers_count,
        language: data.language,
        topics: data.topics || [],
        createdAt: data.created_at,
        updatedAt: data.updated_at,
        openIssues: data.open_issues_count
      };
    } catch (error) {
      throw new Error(`Failed to fetch repository info: ${error}`);
    }
  }

  async getContributors(owner: string, repo: string): Promise<Contributor[]> {
    try {
      const response = await this.client.get(`/repos/${owner}/${repo}/contributors`, {
        params: { per_page: 50 }
      });

      return response.data.map((contributor: any) => ({
        login: contributor.login,
        contributions: contributor.contributions,
        avatarUrl: contributor.avatar_url
      }));
    } catch (error) {
      throw new Error(`Failed to fetch contributors: ${error}`);
    }
  }

  async getLanguages(owner: string, repo: string): Promise<{ [key: string]: number }> {
    try {
      const response = await this.client.get(`/repos/${owner}/${repo}/languages`);
      return response.data;
    } catch (error) {
      throw new Error(`Failed to fetch languages: ${error}`);
    }
  }

  calculateHealth(repo: GitHubRepo, contributors: Contributor[]): RepositoryHealth {
    let score = 0;
    const insights: string[] = [];

    // Star metric (0-25 points)
    if (repo.stars >= 1000) {
      score += 25;
      insights.push('🌟 Excellent community engagement');
    } else if (repo.stars >= 500) {
      score += 20;
      insights.push('⭐ Good community interest');
    } else if (repo.stars >= 100) {
      score += 15;
      insights.push('📈 Growing community');
    } else {
      insights.push('🚀 Building community awareness');
    }

    // Contributor metric (0-25 points)
    if (contributors.length >= 50) {
      score += 25;
      insights.push('👥 Strong contributor base');
    } else if (contributors.length >= 20) {
      score += 20;
      insights.push('🤝 Active community');
    } else if (contributors.length >= 5) {
      score += 15;
      insights.push('👤 Growing team');
    } else {
      insights.push('👤 Building contributor network');
    }

    // Fork metric (0-20 points)
    if (repo.forks >= repo.stars * 0.3) {
      score += 20;
      insights.push('🔀 High fork-to-star ratio');
    } else if (repo.forks >= repo.stars * 0.1) {
      score += 15;
      insights.push('🔀 Moderate fork activity');
    } else {
      score += 10;
    }

    // Issue metric (0-15 points)
    if (repo.openIssues < 20) {
      score += 15;
      insights.push('✅ Well-maintained issue tracker');
    } else if (repo.openIssues < 50) {
      score += 10;
      insights.push('📋 Manageable issue backlog');
    } else {
      insights.push('⚠️ Consider addressing issue backlog');
    }

    // Activity metric (0-15 points)
    const lastUpdate = new Date(repo.updatedAt).getTime();
    const daysSinceUpdate = (Date.now() - lastUpdate) / (1000 * 60 * 60 * 24);
    
    if (daysSinceUpdate < 7) {
      score += 15;
      insights.push('🔥 Actively maintained');
    } else if (daysSinceUpdate < 30) {
      score += 10;
      insights.push('✓ Regularly maintained');
    } else if (daysSinceUpdate < 90) {
      score += 5;
      insights.push('⏱️ Maintenance may be slowing');
    }

    const rating = this.scoreToRating(score);

    return { score, rating, insights };
  }

  private scoreToRating(score: number): 'Excellent' | 'Good' | 'Fair' | 'Needs Improvement' {
    if (score >= 90) return 'Excellent';
    if (score >= 70) return 'Good';
    if (score >= 50) return 'Fair';
    return 'Needs Improvement';
  }

  async analyze(owner: string, repo: string): Promise<AnalysisResult> {
    const startTime = performance.now();

    const [repoInfo, contributors, languages] = await Promise.all([
      this.getRepositoryInfo(owner, repo),
      this.getContributors(owner, repo),
      this.getLanguages(owner, repo)
    ]);

    const health = this.calculateHealth(repoInfo, contributors);
    const analysisTime = Math.round((performance.now() - startTime) * 100) / 100;

    return {
      repository: repoInfo,
      contributors,
      languages,
      health,
      analysisTime
    };
  }
}

// CLI Interface
async function main() {
  const args = process.argv.slice(2);

  if (args.length === 0 || args.includes('--help') || args.includes('-h')) {
    console.log(`
GitHub Analyzer CLI - TypeScript Version
Usage: npx github-analyzer-ts <owner/repo> [options]

Options:
  --token TOKEN    GitHub API token (or set GITHUB_TOKEN environment variable)
  --help, -h       Show this help message

Example:
  npx github-analyzer-ts torvalds/linux
  npx github-analyzer-ts facebook/react --token ghp_xxxxx
    `);
    process.exit(0);
  }

  const repoPath = args[0];
  let token: string | undefined;

  if (args.includes('--token')) {
    const tokenIdx = args.indexOf('--token');
    if (tokenIdx + 1 < args.length) {
      token = args[tokenIdx + 1];
    }
  }

  try {
    const [owner, repo] = repoPath.split('/');
    if (!owner || !repo) {
      throw new Error('Invalid repository format. Use: owner/repo');
    }

    const analyzer = new GitHubAnalyzerTS(token);
    console.log(`\n📊 Analyzing ${owner}/${repo}...\n`);

    const result = await analyzer.analyze(owner, repo);

    // Display results
    console.log('╔════════════════════════════════════════════════════════╗');
    console.log('║           GITHUB REPOSITORY ANALYSIS REPORT            ║');
    console.log('╚════════════════════════════════════════════════════════╝\n');

    console.log(`📦 Repository: ${result.repository.name}`);
    if (result.repository.description) {
      console.log(`📝 Description: ${result.repository.description}`);
    }

    console.log(`\n📊 Statistics:`);
    console.log(`  ⭐ Stars: ${result.repository.stars}`);
    console.log(`  🔀 Forks: ${result.repository.forks}`);
    console.log(`  👁️ Watchers: ${result.repository.watchers}`);
    console.log(`  ⚠️ Open Issues: ${result.repository.openIssues}`);

    console.log(`\n👥 Contributors: ${result.contributors.length}`);
    const topContributors = result.contributors.slice(0, 5);
    topContributors.forEach((c, idx) => {
      console.log(`  ${idx + 1}. ${c.login} (${c.contributions} commits)`);
    });

    console.log(`\n💻 Languages:`);
    Object.entries(result.languages)
      .sort(([, a], [, b]) => b - a)
      .slice(0, 5)
      .forEach(([lang, bytes]) => {
        const percent = ((bytes / Object.values(result.languages).reduce((a, b) => a + b, 0)) * 100).toFixed(1);
        console.log(`  ${lang}: ${percent}%`);
      });

    console.log(`\n🏥 Health Assessment:`);
    console.log(`  Health Score: ${result.health.score}/100`);
    console.log(`  Rating: ${result.health.rating}`);
    console.log(`\n  Insights:`);
    result.health.insights.forEach(insight => {
      console.log(`  ${insight}`);
    });

    console.log(`\n⏱️ Analysis completed in ${result.analysisTime}ms\n`);

  } catch (error) {
    console.error(`❌ Error: ${error instanceof Error ? error.message : String(error)}`);
    process.exit(1);
  }
}

main();

export { GitHubAnalyzerTS, GitHubRepo, Contributor, AnalysisResult };
